src/main/java/com/google/sps/servlets/GroceryServlet.java

```diff
@@ -40,15 +40,25 @@
   public void doPost(HttpServletRequest request, HttpServletResponse response) throws IOException {      
     UserService userService = UserServiceFactory.getUserService();
     DatastoreService datastore = DatastoreServiceFactory.getDatastoreService();
-    String userEmail = userService.getCurrentUser().getEmail();
+    String userEmail;
+    try {
+    userEmail = userService.getCurrentUser().getEmail();
+    } catch (NullPointerException e) {
```

Comment:
Don't catch a NullPointerException here.

Use an if condition to check for null

Reply:
when I tried this I got an nullException error.

Reply:
Can you determine under what cases you are getting a NullPointerException, and why it is happening?

NullPointerExceptions happen when the code tries to dereference something that is null -- which should not be done, and which you can prevent by doing an (if whateverVariableMightBeNull == null) check before dereferencing an object that might be null.

Handling null by catching NullPointerExceptions is dangerous, because it can obscure a bug.

For instance, if `userService` ends up  null here due to a bug (which I assume we expect never ever to happen, since it provides the authentication), the catch block on 46 would catch it and proceed, when we would actually want it to fail (so that the programmer could fix the bug).

Reply:
the only case would be if a user is on the page and isn't signed in

Reply:
in my if statment I have if (userService.getCurrentUser().getEmail().equals(null)) because == gives an error when running and when I go to grocery while not signed in I get a really long error which basically  is telling me there's a null pointer exception instead of the error I want to be shown

Reply:
Are you sure .userService.getCurrentUser().getEmail(). is what you want to check on?

NullPointerExceptions happen when you try to call something on null.

You can break apart the statement (e.g. using printlns) to see what is actually null (userService, userService.getCurrentUser(), userService.getCurrentUser().getEmail()). Then you can do a == check on the thing that is actually null.

Some tips from a webpage explaining that in more detail: https://frequal.com/java/DebuggingNullPointerExceptions.html

Note that .equals can't be used for null checks; .equals is a method and trying to call a method on null results in a NullPointerException. 
== works because it compares the memory addresses (think of how it works in C++).

Reply:
solved